src/basic/copy.c

```diff
@@ -386,14 +386,13 @@ static int fd_copy_directory(
                         continue;
                 }
 
-                if (buf.st_dev != original_device)
-                        continue;
-
                 if (S_ISREG(buf.st_mode))
                         q = fd_copy_regular(dirfd(d), de->d_name, &buf, fdt, de->d_name, override_uid, override_gid, copy_flags);
-                else if (S_ISDIR(buf.st_mode))
+                else if (S_ISDIR(buf.st_mode)) {
+                        if (buf.st_dev != original_device)
+                                continue;
```

Comment:
Thank you. I'm not sure where this `if` should be placed. `fd_copy_directory` stop copying as soon as it detects a filesystem boundary, while `cp --one-file-system` creates an empty directory before stopping copying.